Docs/6a_RuneBasics.md

```diff
@@ -1,5 +1,295 @@
+}
+```
+
+#### JSON
+
+Models require all properties to be `Codable`, so any property that isn't one of the types listed above will be stored as `JSON`.
+
+```swift
+struct Todo: Model {
+    struct TodoMetadata: Codable {
+        var createdAt: Date
+        var lastUpdated: Date
+        var colorName: String
+        var comment: String
+    }
+
+    var id: Int?
+
+    let name: String
+    let isDone: Bool
+    let metadata: TodoMetadata // will be stored as JSON
+}
+```
+
+#### Custom JSON Encoders
+
+By default, `JSON` properties are encoded using a default `JSONEncoder()` and stored in the table column. You can use a custom `JSONEncoder` by overriding the static `Model.jsonEncoder`.
+
+```swift
+struct Todo: Model {
+    static var jsonEncoder: JSONEncoder = {
+        let encoder = JSONEncoder()
+        encoder.outputFormatting = .prettyPrinted
+        return encoder
+    }()
+
+    ...
+}
+```
+
+#### Custom JSON Decoders
+
+Likewise, you can provide a custom `JSONDecoder` for decoding data from JSON columns.
+
+```swift
+struct Todo: Model {
+    static var jsonDecoder: JSONDecoder = {
+        let decoder = JSONDecoder()
+        decoder.dateDecodingStrategy = .iso8601
+        return decoder
+    }()
+
+    ...
+}
+```
+
+## Decoding from `DatabaseRow`
+
+`Model`s may "decoded" from a `DatabaseRow` that was the result of a raw query or query builder query. The `Model`'s properties will be mapped to their relevant columns, factoring in any custom `keyMappingStrategy`. This will throw an error if there is an issue while decoding, such as a missing column.
+
+```swift
+struct User: Model {
+    static let tableName = "users"
+
+    var id: Int?
+
+    let firstName: String
+    let lastName: String
+    let age: String
+}
+
+database.runRawQuery("select * from users")
+    .mapEach { try! $0.decode(User.self) }
+    .whenSuccess { users in
+        for user in users {
+            print("Got user named \(user.firstName) \(user.lastName).")
+        }
+    }
+```
+
+**Note**: For the most part, if you are using Rune you won't need to call `DatabaseRow.decode(_ type:)` because the typed ORM queries described in the next section decode it for you.
+
+## Model Querying
+
+To add some type safety to query builder queries, you can initiate a typed query off of a `Model` with the static `.query` function.
+
+```swift
+let modelQuery: ModelQuery<User> = User.query()
+```
+
+`ModelQuery<M: Model>` is a subclass of the generic `Query`, with a few functions for running and automatically decoding `M` from a query.
+
+### Get All
+
+`.getAll()` returns an EventLoopFuture<[M]> containing all `Model`s that matched the query.
+
+```swift
+User.query()
+    .where("name", in: ["Josh", "Chris", "Rachel"])
+    .getAll() // EventLoopFuture<[User]> of all users named Josh, Chris, or Rachel
+```
+
+### Get First
+
+`.getFirst()` returns an `EventLoopFuture<M?>` containing the first `Model` that matched the query, if it exists.
```

Comment:
Did we have these named differently than the query builder equivalents on purpose? Could see it being a little confusing as to which one to use. When to use `getFirst` vs `first` etc

Reply:
Mm, yeah was thinking about that. The difference is that this one decodes and evaluates eager loads. Maybe I'll just rename this to `decodeFirst()` and `decodeAll()`. Not a huge fan of that though, since it's likely the function that will most likely be called (assuming they're using Rune for most queries).

How about `getRows()` and `firstRow()` for the Query and `get/all()` and `first()` for the `ModelQuery`?